src/impression.js

```diff
@@ -47,15 +85,66 @@ export function maybeTrackImpression(win) {
     // avoid duplicate tracking.
     win.location.hash = '';
   }
+
   viewer.whenFirstVisible().then(() => {
-    invoke(win, clickUrl);
+    // TODO(@zhouyx) need test with a real response.
+    const promise = invoke(win, dev().assertString(clickUrl)).then(response => {
+      applyResponse(win, viewer, response);
+    });
+
+    // Timeout invoke promise after 8s and resolve trackImpressionPromise.
+    timerFor(win).timeoutPromise(TIMEOUT_VALUE, promise,
+        'timeout waiting for ad server response').catch(() => {
+        }).then(() => {
```

Comment:
eh I don't quite get, `resolveImpression()` is a simple resolve function that don't receive any param? you mean make the `trackImpressionPromise` resolve after 8s
```
trackImpressionPromise = new Promise(resolve => {
   const id = timerFor(win).delay(resolve, 8000);
   resolveImpression = () => {
      timerFor(win).cancel(id);
      resolve();
   }
});
```

I don't see this one is much better than the current approach.